### PR TITLE
feat(unittests): add vrl as test input

### PR DESCRIPTION
**Pull request by MichaHoffmann**
Add VRL as a valid input for unittests, this makes it possible to define test inputs as such:

```
[[tests.inputs]]
type = "vrl"
insert_at = "my_transform"
source = '''
. = {
  "message": "foo bar baz",
  "timestamp": now(),
  "tags": {
    "foo": "bar",
    "fizz": "buzz"
  }
}
'''
```

And it kinda works:
```
output payloads from ["my_transform"] (events encoded as JSON):
  {"message":"foo bar baz","tags":{"fizz":"buzz","foo":"bar"},"timestamp":"2023-11-09T20:17:08.072518248Z"}
```

Since i basically just hacked around a bit; is this something that is useful and worthy of refining?

Inspired by #12490 and #15304

Closes: https://github.com/vectordotdev/vector/issues/15304

### CODE REVIEW

File: src/config/mod.rs
Comment (pront) [inline src/config/mod.rs:508]: Unrelate to this PR but `TestInput` should have been an enum.

**Review by pront — APPROVED** [review]
Hi @MichaHoffmann, thanks for the PR. LGTM.

**Comment by pront** [comment]
Let's wait for someone from the core team to also take a look before merging.

**Comment by MichaHoffmann** [comment]
Hey @pront; do i have to request a review from someone from the core team?

File: src/config/unit_test/mod.rs
Comment (bruceg) [inline src/config/unit_test/mod.rs:571]: Using `map_err` makes this a little more concise:
```suggestion
                let result = vrl::compiler::compile(source, &fns)
                    .map_err(|e| Formatter::new(source, e.clone()).to_string())?;
```

**Review by bruceg — APPROVED** [review]
LGTM

**Comment by MichaHoffmann** [comment]
@bruceg ~ that performance regression merge queue problem cannot be related to this change i suppose; right?

**Comment by pront** [comment]
> @bruceg ~ that performance regression merge queue problem cannot be related to this change i suppose; right?

Looks unrelated.

File: src/config/unit_test/mod.rs
Comment (jszwedko) [inline src/config/unit_test/mod.rs:583]: It would be nice to expand this to metrics in the future, but this is a good start since unit tests are frequently used for logs. We should make this limitation clear in the docs though.

**Review by jszwedko — CHANGES_REQUESTED** [review]
This is very useful, thank you @MichaHoffmann !

I think one thing that is missing is updating the documentation to describe this new behavior: https://github.com/vectordotdev/vector/edit/master/website/content/en/docs/reference/configuration/unit-tests.md

File: src/config/unit_test/mod.rs
Comment (MichaHoffmann) [reply src/config/unit_test/mod.rs:583]: Thats more me being nwe to the code base i suppose! Since the value is just a blob we could probably also parse a metric event from it if the schema matches; i just didnt know how.

File: src/config/unit_test/mod.rs
Comment (MichaHoffmann) [reply src/config/unit_test/mod.rs:571]: Did that now! Thanks

File: website/content/en/docs/reference/configuration/unit-tests.md
Comment (alai97) [inline website/content/en/docs/reference/configuration/unit-tests.md:224]:
```suggestion
`type` | string | The type of input you're providing. [`vrl`](#logs), [`log`](#logs), [`raw`](#logs), or [`metric`](#metrics) are currently the only valid values.
```

File: website/content/en/docs/reference/configuration/unit-tests.md
Comment (alai97) [inline website/content/en/docs/reference/configuration/unit-tests.md:229]:
```suggestion
`source` | string (vrl program) | If the transform handles [log events](#logs), the result of the vrl program will be the input event.
```

**Review by alai97 — COMMENTED** [review]
Looks good, two suggestions!

**Review by bruceg — APPROVED** [review]

File: website/content/en/docs/reference/configuration/unit-tests.md
Comment (MichaHoffmann) [reply website/content/en/docs/reference/configuration/unit-tests.md]: Thanks!

File: website/content/en/docs/reference/configuration/unit-tests.md
Comment (MichaHoffmann) [reply website/content/en/docs/reference/configuration/unit-tests.md]: Thanks!

**Review by bruceg — APPROVED** [review]

**Comment by Salathaniel** [comment]
https://gptzero.me/?via=salathaniel

**Pull request merged**
